tooling/greybox_fuzzer/src/coverage.rs

```diff
@@ -0,0 +1,751 @@
+        let all_witnesses_in_bool_coverage: HashSet<_> = new_coverage
+            .acir_bool_coverage
+            .iter()
+            .map(|acir_bool_state| acir_bool_state.witness_id)
+            .collect();
+
+        let mut states_to_remove = Vec::new();
+        // Check that all boolean state witnesses observed in accumulated coverage are booleans here, too
+        for state in self.acir_bool_coverage.iter() {
+            if !all_witnesses_in_bool_coverage.contains(&state.witness_id) {
+                states_to_remove.push(*state);
+            }
+        }
+
+        // Remove states that are not booleans
+        for state in states_to_remove {
+            self.acir_bool_coverage.remove(&state);
+            add_to_leavers(Some(self.bool_state_to_testcase_id[&state]));
+        }
+    }
+
+    /// Returns true if there is new coverage in the presented testcase
+    pub fn detect_new_coverage(&self, new_coverage: &SingleTestCaseCoverage) -> bool {
+        // Go through all single branch coverage ranges and check that either:
+        // 1. A new branch is taken
+        // 2. A branch is taken more times than ever encountered before in a single execution
+        // 3. A branch is taken pow2 times that hasn't been previously observed
+        for branch in self.brillig_branch_coverage.iter() {
+            let testcase_value = new_coverage.brillig_coverage[branch.raw_index];
+            if !testcase_value.is_zero() {
+                if (branch.encountered_loop_log2s
+                    | 1u32
+                        << (if testcase_value.is_zero() { 0 } else { testcase_value.ilog2() + 1 }))
+                    != branch.encountered_loop_log2s
+                {
+                    return true;
+                }
+                if testcase_value > branch.encountered_loop_maximum {
+                    return true;
+                }
+            }
+        }
+        // Go through comparison coverage and detect:
+        // 1. If a particular comparison has achieved a difference between arguments whose log2 is smaller than previously observed
+        // 2. If the smallest log2 previously observed has been detected more times in the same execution
+        // 3. If the number of executions of the log2 is a new log2 that hasn't been observed
+        for cmp_approach in self.brillig_cmp_approach_coverage.iter() {
+            if !cmp_approach.enabled {
+                // No need to detect closeness any more if we've hit the equality case
+                continue;
+            }
+            let mut least_different_bits = u32::MAX;
+            let mut last_value = 0;
+            for i in 0..(cmp_approach.bits + 1) {
+                if !new_coverage.brillig_coverage[i + cmp_approach.raw_index].is_zero() {
+                    least_different_bits = (cmp_approach.bits - i) as u32;
+                    last_value = new_coverage.brillig_coverage[i + cmp_approach.raw_index];
+                }
+            }
```

Comment:
Could this use `find_closest_comparison`?

Reply:
Updated